### DISCUSSION TITLE
XIAO ESP32S3 channel values not updating

### DISCUSSION THREAD

**Discussion by matt-nolan11** [ZZ-Cat/CRSFforArduino#54] [Q&A] [answered]
Hello!
First, I really appreciate and admire this project, and I'm excited that the first major release is coming soon.

I'm running into an issue using the "channels" example on a Seeeduino XIAO ESP32-S3, using a Radiomaster TX16S and a RM RP2 ELRS receiver. The example start up ok, but the channel values do not update with my radio's stick movements. They seem stuck at what I assume are default channel values (see the below screenshot)
![image](https://github.com/ZZ-Cat/CRSFforArduino/assets/46466472/915a6d73-fd72-4429-b012-36fb2823cf64)
I'm able to run the same example code on a XIAO SAMD21, so I'm pretty sure my radio hardware is set up correctly.

I think the issue might be related to the ESP32 using Serial0 for USB communications with the PC. There are two other UARTs available through pin multiplexing, but I think CRSFfor Arduino does not support that functionality right now. I do note that my board is listed as supported, however. Do you know of any fixes, or should I just wait for UART pin assignment to become a thing?

**Comment by ZZ-Cat** [comment]
Yea, ESP32 chipsets isn't something I have available on hand, so I am relying on feedback from others to let me know whether-or-not something is working, and to help me with testing stuff out on their hardware.

UART Pin assignments are in the cards for version 1.0.0, as it's currently not implemented... yet.

For now, CRSF for Arduino defaults to Serial1 on every target. So, not entirely sure why Serial0 would interfere with that. But... stranger things have happened.

How does the Channels example fair on your Xiao SAMD21?

**Reply by matt-nolan11** [reply]
Interesting, sounds like it's probably not related to Serial0 then. As far as I know Serial1 is not available for use on a lot of ESP boards without pin remapping, so maybe the example is properly reading to the default Serial1 MCU pins, which are simply not user-accessible.

The Channels example seems to work perfectly on the Xiao SAMD21. I've tested the telemetry example as well, and it also works. Unfortunately for me, there's another library I'd like to use for my project, and it only supports ESP32, so it looks like I'm stuck for now.

**Reply by ZZ-Cat** [reply]
What is the default UART port for ESP32?
In the meantime, I _can_ change it to that for the time being, if need be.

**Reply by matt-nolan11** [reply]
Most full-sized dev kits have Serial0 and Serial2 pinned out. Unfortunately, the smaller Xiao only has Serial0 exposed, so that might complicate things. It's a questionable choice by Seeed imo, but there it is.

**Reply by ZZ-Cat** [reply]
Well, I will see what I can do.  
For now, sit tight, 'cause I'm on hiatus right now. I won't be getting back into developing this again until the 14th of Januar, next year.

**Reply by matt-nolan11** [reply]
Sounds good, thanks for your help so far! There's a lot of other stuff I need to work on for this project, so I'll have plenty to do in the meantime :)

**Reply by ZZ-Cat** [reply, marked as answer]
Okay, according to my information, the Seeed XIAO series all use pins D6 and D7 for their default UART.  
This is different to the more commonplace D0 and D1 pins.

D6 is Tx and D7 is Rx.

The current version of CRSF for Arduino _only_ uses pins D0 and D1.  
In the upcoming version (Currently being tracked in #61), not only will I give you the ability to use arbitrary UART ports, I _may_ take the liberty of re-factoring the default UART pin assignments for _all_ Seeed XIAO targets.  

It also appears to me that all of Adafruit's QtPy variants get even _more_ wild with their UART pin assignments.  
'Cause I originally went off the assumption that everyone is keeping to the D0 and D1 assignments that Arduino themselves originally started up... _boy was I mistaken!_

So, whatever they all default to, I _may_ need to put that into the development of Version 1.0.0.

Anyway... to get that into the pipeline, can you put this as an Issue under the [Bug Report](https://github.com/ZZ-Cat/CRSFforArduino/issues/new?assignees=&labels=%F0%9F%90%9E+Bug+%F0%9F%90%9E%2CPending+%F0%9F%93%8C&projects=&template=bug-report.yml) category? Thanks in advance.

**Reply by matt-nolan11** [reply]
Thanks for the answer! I can still submit the bug report if you want, but I have some new results to share after downloading the most recent branch:

It looks like I can use whichever pins I like by using the HardwareSerial.setPins() method at the very start of setup()
One important thing to note is that you must use the "GPIO" pin numbers rather than the "D" pin numbers. For my board:

![image](https://github.com/ZZ-Cat/CRSFforArduino/assets/46466472/b479c74e-efd7-4f99-8da3-18cab5208be9)
[Source](https://wiki.seeedstudio.com/xiao_esp32s3_pin_multiplexing/)

I started with the channels.ino example, and added the line "Serial1.setPins(44, 43);" right at the start of setup() in order to use pins D7 and D6. Everything works as expected! I tried the same thing with GPIO7 and GPIO6 (pins D8 and D5), and this worked as well.

**Reply by ZZ-Cat** [reply]
> Thanks for the answer! I can still submit the bug report if you want, but I have some new results to share after downloading the most recent branch:
> 
> It looks like I can use whichever pins I like by using the HardwareSerial.setPins() method at the very start of setup()
> One important thing to note is that you must use the "GPIO" pin numbers rather than the "D" pin numbers. For my board:
> 
> ![image](https://github.com/ZZ-Cat/CRSFforArduino/assets/46466472/b479c74e-efd7-4f99-8da3-18cab5208be9)
> [Source](https://wiki.seeedstudio.com/xiao_esp32s3_pin_multiplexing/)
> 
> I started with the channels.ino example, and added the line "Serial1.setPins(44, 43);" right at the start of setup() in order to use pins D7 and D6. Everything works as expected! I tried the same thing with GPIO7 and GPIO6 (pins D8 and D5), and this worked as well.
> 
> 

Excellent! No bug report needed, 'cause it looks like I fixed it when I brought the  `HardwareSerial` class up to the Sketch Layer.

I'm glad to hear it's working now. =^/.^=
